pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go

```diff
@@ -75,26 +75,30 @@ func mergeTransportSocketCluster(c *cluster.Cluster, cp *model.EnvoyFilterConfig
 
 	var tsmPatch *core.TransportSocket
 
-	// Test if the patch contains a config for TransportSocket
-	// and if the cluster contains a config for Transport Socket Matches
-	if cpValueCast.GetTransportSocket() != nil && c.GetTransportSocketMatches() != nil {
-		for _, tsm := range c.GetTransportSocketMatches() {
-			if tsm.GetTransportSocket() != nil && cpValueCast.GetTransportSocket().Name == tsm.GetTransportSocket().Name {
-				tsmPatch = tsm.GetTransportSocket()
-				break
+	// Check if cluster patch has a transport socket.
+	if cpValueCast.GetTransportSocket() != nil {
+		// First check if the transport socket matches with any cluster transport socket matches.
+		if len(c.GetTransportSocketMatches()) > 0 {
+			for _, tsm := range c.GetTransportSocketMatches() {
+				if tsm.GetTransportSocket() != nil && cpValueCast.GetTransportSocket().Name == tsm.GetTransportSocket().Name {
+					tsmPatch = tsm.GetTransportSocket()
+					break
+				}
+			}
+			if tsmPatch == nil && len(c.GetTransportSocketMatches()) > 0 {
+				// If we merged we would get both a transport_socket and transport_socket_matches which is not valid
+				// Drop the filter, but indicate that we handled the merge so that the outer function does not try
+				// to merge it again
+				return true, nil
+			}
+		} else if c.GetTransportSocket() != nil {
+			if cpValueCast.GetTransportSocket().Name == c.GetTransportSocket().Name {
+				tsmPatch = c.GetTransportSocket()
 			}
 		}
-		if tsmPatch == nil && len(c.GetTransportSocketMatches()) > 0 {
-			// If we merged we would get both a transport_socket and transport_socket_matches which is not valid
-			// Drop the filter, but indicate that we handled the merge so that the outer function does not try
-			// to merge it again
-			return true, nil
-		}
-	} else if cpValueCast.GetTransportSocket() != nil && c.GetTransportSocket() != nil {
-		if cpValueCast.GetTransportSocket().Name == c.GetTransportSocket().Name {
-			tsmPatch = c.GetTransportSocket()
-		} else {
-			// There is a name mismatch, so we cannot do a deep merge. Instead just replace the transport socket
+		// This means either there is a name mismatch or cluster does not have transport socket matches/transport socket.
+		// We cannot do a deep merge. Instead just replace the transport socket
+		if tsmPatch == nil {
```

Comment:
Isn't this shoube under `else if c.GetTransportSocket() != nil`

Reply:
It really does not matter. But refactored a bit more to simplify further. PTAL.